### PR TITLE
feat(ext-plugin-post-resp): support get response body by extra_info

### PR DESCRIPTION
**Pull request by soulbird**
### Description

Provide the `ext-plugin-post-resp` plugin with the ability to get the response body, so that the External Plugin will be able to do some processing on the response body.
As an additional feature, the `ext-plugin-post-resp` plugin also supports getting the request context in the response phase.

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: t/plugin/ext-plugin/extra-info.t [hunk 1]
```diff
@@ -181,3 +182,161 @@ GET /hello
...
+
+        content_by_lua_block {
+            local ext = require("lib.ext-plugin")
+            local actions = {
+                {type = "respbody", result = nil}
+            }
+            ext.go({extra_info = actions})
+        }
+    }
+--- error_code: 200
```
Comment (tzssangglass) [inline t/plugin/ext-plugin/extra-info.t:203]: would valid error_code by  default?

File: t/plugin/ext-plugin/extra-info.t [hunk 2]
```diff
@@ -181,3 +182,161 @@ GET /hello
...
+--- grep_error_log eval
+qr/send extra info req successfully/
+--- grep_error_log_out
+send extra info req successfully
+--- response_body
+hello world
+
+
+
+=== TEST 10: ask request body (not exist)
```
Comment (tzssangglass) [inline t/plugin/ext-plugin/extra-info.t:327]: why we check request body here?

Comment (soulbird) [reply t/plugin/ext-plugin/extra-info.t:327]: I reused the [handle_extra_info](https://github.com/apache/apisix/blob/master/apisix/plugins/ext-plugin/init.lua#L283) function to get the response body, the request body is part of this function.

File: t/plugin/ext-plugin/extra-info.t (hunk 1 above)
Comment (soulbird) [reply t/plugin/ext-plugin/extra-info.t:203]: Sure, I'll remove it.

**Review by tzssangglass — APPROVED** [review]

**Comment by tzssangglass** [comment]
LGTM

File: apisix/plugins/ext-plugin/init.lua [hunk 3]
```diff
@@ -304,7 +305,31 @@ local function handle_extra_info(ctx, input)
...
+            local info = req:Info()
+            local respbody_req = extra_info_respbody.New()
+            respbody_req:Init(info.byte, info.pos)
+
+            local chunks = {}
+            local err = helper.response_reader(ext_res.body_reader, function (chunk, chunks)
+                -- When the upstream response is chunked type,
+                -- we will receive the complete response body
+                -- before sending it to the runner program
+                -- to reduce the number of RCP calls.
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/init.lua:320]:
```suggestion
                -- to reduce the number of RPC calls.
```

File: t/plugin/ext-plugin/extra-info.t [hunk 4]
```diff
@@ -20,6 +20,7 @@ repeat_each(1);
 no_long_string();
 no_root_location();
 no_shuffle();
+log_level("info");
```
Comment (spacewander) [inline t/plugin/ext-plugin/extra-info.t:23]: The default level is already info? https://github.com/apache/apisix/blob/a47d05a9a32db1ba7de6bda68c67cf0008f8442c/t/APISIX.pm#L24

File: t/plugin/ext-plugin/extra-info.t [hunk 5]
```diff
@@ -181,3 +182,156 @@ GET /hello
 --- error_code: 503
 --- error_log
 failed to receive RPC_HTTP_REQ_CALL: closed
+
+
+
+=== TEST 5: ask response body (not exist)
```
Comment (spacewander) [inline t/plugin/ext-plugin/extra-info.t:188]: Let's add a test with empty response body

File: apisix/plugins/ext-plugin/helper.lua [hunk 6]
```diff
@@ -56,4 +56,27 @@ function _M.get_conf_token_cache_time()
...
 
 
+function _M.response_reader(reader, callback, ...)
+    if not reader then
+        return "get response reader failed"
+    end
+
+    repeat
+        local chunk, read_err, cb_err
+        -- TODO: HEAD or 304
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/helper.lua:66]: Could you use at least one sentence for this TODO? I can't recall why this comment is added.

File: t/plugin/ext-plugin/extra-info.t (hunk 5 above)
Comment (soulbird) [reply t/plugin/ext-plugin/extra-info.t:188]: It doesn't make sense to add a test with an empty response body in the `ext-plugin-pre-req` plugin. I guess you want me to add this test case in the `ext-plugin-post-resp` plugin.
However, an empty response body usually occurs when the upstream returns a 304 status code or the request method is HEAD, which needs to be handled separately in subsequent PRs. I have added `TODO`.

File: apisix/plugins/ext-plugin/helper.lua (hunk 6 above)
Comment (soulbird) [reply apisix/plugins/ext-plugin/helper.lua]: done

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]
